### PR TITLE
Enables torchrun for XLA-based accelerators

### PR DESCRIPTION
**Pull request by jeffhataws**
# What does this PR do?

This PR enables torchrun for XLA-based accelerators (TPU/NeuronCore) by using torch.distributed XLA backend. It is dependent on the torch/xla change https://github.com/pytorch/xla/pull/3609.

Example application is the AWS Neuron tutorial with HF Trainer that uses torchrun:

https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/tutorials/training/finetune_hftrainer.html

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1325,6 +1325,10 @@ def _wrap_model(self, model, training=True, dataloader=None):
         if not training:
             return model
 
+        # add torchrun support for XLA, but currently DDP doesn't support XLA yet
+        if is_torch_tpu_available():
+            return model
+
```
Comment (sgugger) [inline src/transformers/trainer.py:1331]: No the model needs to be wrapped in a `DistributedDataParallel` for TPU,

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -53,14 +53,16 @@
...
     import torch
     import torch.distributed as dist
 
 if is_torch_tpu_available(check_device=False):
     import torch_xla.core.xla_model as xm
-
+    # torchrun support
+    import torch_xla.distributed.xla_backend
+    if os.environ.get("WORLD_SIZE"):
+        torch.distributed.init_process_group(backend="xla")
```
Comment (sgugger) [inline src/transformers/training_args.py:65]: This should be moved to the `_setup_devices` method of `TrainingArguments`. It shouldn't be done on import.

File: tests/trainer/test_trainer_distributed.py [hunk 3]
```diff
@@ -61,6 +62,22 @@ def forward(self, input_ids, labels=None):
             else:
                 return input_ids
 
+class TestTrainerDistributedXLA(TestCasePlus):
+    @require_torch_tpu
```
Comment (sgugger) [inline tests/trainer/test_trainer_distributed.py:66]: We don't have any runner testing on XLA for now, so this test will never be run FYI.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! There are a few problems with it though, as it breaks current support for PyTorch XLA.

File: src/transformers/trainer.py (hunk 1 above)
Comment (jeffhataws) [reply src/transformers/trainer.py]: Thanks. Limited the case to torchrun and clarify that currently DDP doesn't work with torch.distributed XLA backend yet (https://github.com/pytorch/pytorch/issues/79164)

File: tests/trainer/test_trainer_distributed.py (hunk 3 above)
Comment (jeffhataws) [reply tests/trainer/test_trainer_distributed.py:66]: Thanks. Will get help with running these tests.

File: src/transformers/training_args.py (hunk 2 above)
Comment (jeffhataws) [reply src/transformers/training_args.py:65]: Thanks. Unfortunately I can't move it into _setup_devices for some reason as it run into device not found error. Will investigate.

File: src/transformers/trainer.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/trainer.py]: In this case we should wait until it's supported. This will break the current TPU support.

**Pull request closed without merging**